### ISSUE TITLE
os/platform-specific fields in stack.yaml

### ISSUE THREAD

**Issue by leohaskell** [commercialhaskell/stack#3369]
I think it would be nice to have a way of specifying fields that are os-specific. Something like this:

```yaml
os-specific:
  windows:
    ghc-variant: integersimple
    flags:
      hashable:
        integer-gmp: false
      sceintific:
        integer-simple: true
...
```
Thus, one can keep more build information within stack and write less additional scripts.

NOTE: This could be possible also by modifying the confHook in Setup.hs if stack didn't bypass the configure step.

**Comment by mgsloan** [comment]
Potentially related issue: https://github.com/commercialhaskell/stack/issues/2048

Adding conditionals could be a bit complicated, might be worth it though.  I don't have a strong feeling about it, but historically we've avoided making the configuration format very fancy.

**Comment by lwm** [comment]
I would mark this as a duplicate of #2048, no?

https://github.com/commercialhaskell/stack/issues/2048#issuecomment-212161703 seems to be the best solution for this so far.

**Comment by saurabhnanda** [comment]
Is there a way to specify platform specific stack **overrides** in my home folder, i.e. outside of the project root?

I'm on mac osx and need different cabal flags for hlibsass, while everyone else on my team is on linux and needs different flags.

**Comment by leohaskell** [comment]
If you need just different cabal flags, you can pass them like this:
```stack build --flag hlibsass:externallibsass```

**Comment by dbaynard** [comment]
To be clear, this is flags for dependencies, correct?

For a package itself one can use the [os conditional](https://www.haskell.org/cabal/users-guide/developing-packages.html#conditions) in the cabal/hpack file.

One can set flags in a machine's `config.yaml`, rather than the project's `stack.yaml`. Would this work for your use case?

**Comment by pwaller** [comment]
> To be clear, this is flags for dependencies, correct?

Yes. The packages whose flags need to be influenced are not under my control.

> One can set flags in a machine's `config.yaml`, rather than the project's `stack.yaml`. Would this work for your use case?

That's not great either, since it is an additional step that must be documented and remembered to be done during setup of the machine. It would be much better if it were simply a property of the stack.

**Comment by dbaynard** [comment]
I can see good reasons to have such tooling. In one sense it is similar to haddock — often there are haddock build failures with dependencies, but stack currently has no way to indicate that a build should ignore these (the stackage-curator tool does, however).

This broadens the proposal slightly — perhaps it would be better to invert the tree:

```
ghc-variant:
  windows: integersimple
flags:
  hashable:
    windows:
      integer-gmp: false
  scientific [sic]:
    windows:
      integer-simple: true
```

That way any key (within some subset of the top level keys, perhaps) can have such conditionality.

---

For the examples I've seen, however, it seems the best case is actually to fix these issues upstream. For example, while it would be very useful to have the haddock workaround, the real fix would be for upstream libraries to fix their docs.

Likewise,

> …we build llvm-hs with `-f-shared-llvm`, except on Darwin, where this is not currently supported.

Surely this should be fixed by `llvm-hs` setting that flag to false? And if they can't do that, you could fork the package, change the cabal file, and use a custom snapshot which uses your fork?
